### PR TITLE
gh-113939: Frame clear, clear locals

### LINKED ISSUE #113939: `traceback.clear_frames` does not clear locals when there have been previous access to `f_locals`
# Bug report

### Bug description:

```python
import traceback
import gc


class Obj:
    def __init__(self, name: str):
        self.name = name

    def __repr__(self):
        return f"Obj({self.name!r})"

    def __del__(self):
        print("del", self)


def deep(i: int):
    a = Obj(f"a, i={i}")
    if i == 2:
        raise Exception(f"exception at i={i}")
    print(a)


def func():
    for i in range(5):
        gc.collect()
        print("** i:", i)
        try:
            deep(i)
        except Exception as exc:
            print("caught", exc)
            print_tb(exc.__traceback__)
            # traceback.clear_frames(prev_exc.__traceback__)
            clear_tb(exc.__traceback__)
            continue  # continue with next i
        print("deep", i, "done")


def print_tb(tb):
    print("Call stack:")
    while tb:
        frame_i = tb.tb_frame.f_locals.get("i")
        print(f"  {tb.tb_frame.f_code.co_name}: i={frame_i}")
        tb = tb.tb_next


def clear_tb(tb):
    print("Clearing stack:")
    while tb:
        print(tb.tb_frame)
        try:
            tb.tb_frame.clear()
        except RuntimeError:
            print("  cannot clear?")
        else:
            print("  cleared")
            # Using this code triggers that the ref actually goes out of scope, otherwise it does not!
            # print("  now:", tb.tb_frame.f_locals)
        tb = tb.tb_next


if __name__ == '__main__':
    func()
    print("exit")
```

Running this code gives the following output:
```
** i: 0
Obj('a, i=0')
del Obj('a, i=0')
deep 0 done
** i: 1
Obj('a, i=1')
del Obj('a, i=1')
deep 1 done
** i: 2
caught exception at i=2
Call stack:
  func: i=2
  deep: i=2
Clearing stack:
<frame at 0x7f9ee1cc72a0, file '/u/zeyer/code/playground/py-oom-out-of-scope.py', line 34, code func>
  cannot clear?
<frame at 0x7f9ee1c168c0, file '/u/zeyer/code/playground/py-oom-out-of-scope.py', line 20, code deep>
  cleared
** i: 3
Obj('a, i=3')
del Obj('a, i=3')
deep 3 done
** i: 4
Obj('a, i=4')
del Obj('a, i=4')
deep 4 done
exit
del Obj('a, i=2')
```

You see that `Obj('a, i=2')` only is deleted at exit.

This only happens when the `print_tb` is used before, which will access `f_locals` of each frame.

`traceback.clear_frames` should have cleared the locals. But as you see from the output, it does not.

`clear_tb` is basically a copy of `traceback.clear_frames`.

The problem goes away if you access `tb.tb_frame.f_locals` *after it was cleared* (i.e. `tb.tb_frame.clear()` was called).

Looking at the C code, this is what `tb_frame.clear()` will do:
https://github.com/python/cpython/blob/3.12/Objects/frameobject.c#L933-L946
```
static int
frame_tp_clear(PyFrameObject *f)
{
    Py_CLEAR(f->f_trace);

    /* locals and stack */
    PyObject **locals = _PyFrame_GetLocalsArray(f->f_frame);
    assert(f->f_frame->stacktop >= 0);
    for (int i = 0; i < f->f_frame->stacktop; i++) {
        Py_CLEAR(locals[i]);
    }
    f->f_frame->stacktop = 0;
    return 0;
}
```

However, if you accessed `tb_frame.f_locals` before, it will have created a dictionary in `frame->f_locals` here: https://github.com/python/cpython/blob/5c238225f60c33cf1931b1a8c9a3310192c716ae/Objects/frameobject.c#L1218C18-L1218C33

That `frame->f_locals` dict will also have references to all the local vars. And that `f_locals` dict is not cleared in `tb_frame.clear()`.

However, then when you access `tb_frame.f_locals` again, it will update the existing `frame->f_locals` dict, and delete all the local vars in it, because they are not available anymore. Here:
https://github.com/python/cpython/blob/3.12/Objects/frameobject.c#L1256C13-L1256C55

I think it's a bug (or at least very unexpected) that `tb_frame.clear()` does not clear `frame->f_locals`.

So my suggestion would be to add `Py_CLEAR(f->f_frame->f_locals)` in `frame_tp_clear`.

---

There is then another related issue: When the `except` block is left, the exception goes out of scope, so then it should free all the locals (even when `frame.clear()` was not called). However, this is also not the case.

After inspecting this further: Once `frame.f_locals` was accessed from the current frame where the exception is handled, this `frame.f_locals` still has a reference to the exception, and thus to the frames, even though the `DELETE_FAST` for the exception deleted it from the fast locals. See the comments below for more on this.

---

Note, for PyTorch and others, when you first do extended exception reporting which accesses `f_locals` in any way, this here fixes two arising problems. Related:
* https://github.com/pytorch/pytorch/issues/18853
* https://github.com/pytorch/pytorch/issues/27600

E.g., this came up for us because we have this extended exception reporting, which accesses `f_locals`:
```python
        # Extend exception message by module call stack.
        module_names_by_id = {}  # id -> name
        for name, mod in model.named_modules():
            if id(mod) not in module_names_by_id:
                module_names_by_id[id(mod)] = name or "(root)"
        exc_ext = []
        for frame in iter_traceback(exc.__traceback__):
            if frame.f_code.co_nlocals == 0:
                continue
            frame_self = frame.f_locals.get("self")
            if isinstance(frame_self, (torch.nn.Module, rf.Module)):
                func = get_func_from_code_object(frame.f_code, frame=frame)
                if func and func.__name__ and func.__name__.startswith("_") and not func.__name__.startswith("__"):
                    continue
                func_name = (func and func.__qualname__) or type(frame_self).__name__
                exc_ext.append(f"({func_name}) {module_names_by_id.get(id(frame_self), '(unknown)')}")
        if not exc_ext:
            exc_ext.append("(No module call frames.)")
        if len(exc.args) == 1 and isinstance(exc.args[0], str) and not always_direct_print:
            exc.args = ("\n".join([exc.args[0], "", "Module call stack:"] + exc_ext),)
        else:
            print("Module call stack:", file=log.v3)
            for msg in exc_ext:
                print(msg, file=log.v3)
```
The normal `traceback.clear_frames` here does not help.

---

### CPython versions tested on:

3.11, 3.12, 3.13

### Operating systems tested on:

Linux

### Linked PRs
* gh-113940

### PR DESCRIPTION
**Pull request by albertz**
`frame.f_locals` might still contain references to the local vars.

Fix #113939.

Note, for PyTorch and others, when you first do extended exception reporting which accesses `f_locals` in any way, this here fixes two arising problems. Related:
* https://github.com/pytorch/pytorch/issues/18853
* https://github.com/pytorch/pytorch/issues/27600

---

* Issue: gh-113939

### CODE REVIEW

**Comment by albertz** [comment]
> Most changes to Python [require a NEWS entry](https://devguide.python.org/committing/#updating-news-and-what-s-new-in-python). Add one using the [blurb_it](https://blurb-it.herokuapp.com) web app or the [blurb](https://pypi.org/project/blurb/) command-line tool.

I'm not really sure what to write into the NEWS file about this. Suggestions? Maybe you can just add this yourself.

File: Misc/NEWS.d/next/Core and Builtins/2024-01-12-16-40-07.gh-issue-113939.Yi3L-e.rst [hunk 1]
```diff
@@ -0,0 +1,7 @@
+frame.clear(), clear locals, not just fast locals. This is relevant once
```
Comment (iritkatriel) [inline Misc/NEWS.d/next/Core and Builtins/2024-01-12-16-40-07.gh-issue-113939.Yi3L-e.rst]: Can you reword this to be a clear description of what the change in this PR is?  The first sentence in particular is hard to understand.

File: Lib/test/test_frame.py [hunk 2]
```diff
@@ -442,5 +461,28 @@ def dummy():
         # The following line should not cause a segmentation fault.
         self.assertIsNone(frame.f_back)
 
+class LocalsTest(unittest.TestCase):
+    """
+    Tests for locals.
+    """
+
+    def test_locals_cur_frame_issue113939(self):
```
Comment (iritkatriel) [inline Lib/test/test_frame.py]: Ideally the name of a test should be more descriptive (and the issue in a comment):

```suggestion
    def test_locals_cleared_after_exception_handled(self):
        # see gh-113939
```

File: Lib/test/test_frame.py [hunk 3]
```diff
@@ -442,5 +461,28 @@ def dummy():
...
+            nonlocal wr
+            c = C()
+            wr = weakref.ref(c)
+            1/0
+        try:
+            inner()
+        except ZeroDivisionError as exc:
+            support.gc_collect()
+            self.assertIsNotNone(wr())
+            print(exc.__traceback__.tb_frame.f_locals)
```
Comment (iritkatriel) [inline Lib/test/test_frame.py]: print?

File: Python/bytecodes.c [hunk 4]
```diff
@@ -1508,6 +1508,7 @@ dummy_func(
             PyObject *v = GETLOCAL(oparg);
             ERROR_IF(v == NULL, unbound_local_error);
             SETLOCAL(oparg, NULL);
+            Py_CLEAR(LOCALS());
```
Comment (markshannon) [inline Python/bytecodes.c]: What's the reason for this? `DELETE_FAST` should delete a single "fast" local variable, which should have nothing to do with `f_locals`.

**Review by markshannon — COMMENTED** [review]
The change to `DELETE_FAST` seems odd to me.

File: Python/bytecodes.c (hunk 4 above)
Comment (albertz) [reply Python/bytecodes.c]: The rational is explained in the issue #113939. To summarize: I think you would expect that the exception object (and its linked traceback, the frames, and locals of the frames) would go out of scope when you leave the `except` block, and then deleted, when there are no other references to it. The `DELETE_FAST` then would delete the object (ref count becomes 0). However, if `f_locals` was accessed at some point, a `DELETE_FAST` would not trigger the delete, as another copy of the locals is in `f_locals`. It would stay alive as long as you access `f_locals` again which would resync the locals, or once the function leaves.

My use case was a training loop, where the function is running for a very long time, and the locals of some frames where an exception might have occured contained huge amount of memory, and it was very unexpected to me, aber making sure every reference to the exception was deleted, that the memory still was not deleted. Even more unexpectedly was, after getting another exception, then the memory was freed. Only then to find out that this `f_locals` dict was still holding a reference, and once I accessed `f_locals` again (in the next exception later), it was freed.

So, this change here is one possible simple fix, to get the expected behavior. But I understand that this is maybe not how you want it. The question is, how do you want it? Do you have a better suggestion? Or is this all expected behavior, and no change should be done here? (To me, it was unexpected behavior.)

File: Lib/test/test_frame.py (hunk 3 above)
Comment (albertz) [reply Lib/test/test_frame.py]: I actually just want to access `exc.__traceback__.tb_frame.f_locals` here (trigger a `getattr`), which is important for the test. I just thought that also printing the result might be interesting for debugging purpose, but we could also leave away the print.

File: Python/bytecodes.c (hunk 4 above)
Comment (albertz) [reply Python/bytecodes.c]: The test `test_locals_cleared_after_exception_handled` tests exactly this.

File: Misc/NEWS.d/next/Core and Builtins/2024-01-12-16-40-07.gh-issue-113939.Yi3L-e.rst (hunk 1 above)
Comment (albertz) [reply Misc/NEWS.d/next/Core and Builtins/2024-01-12-16-40-07.gh-issue-113939.Yi3L-e.rst]: I reworded this.

File: Python/bytecodes.c (hunk 4 above)
Comment (iritkatriel) [reply Python/bytecodes.c]: Since this is a particular problem related to the lifetime of a handled exception, I think we should think of this as a problem with exceptions, and try to fix it within the exception handling mechanism rather than change `DELETE_FAST`.  

Can we make POP_EXCEPT clear the locals, and move it to after the 

```
               LOAD_CONST               0 (None)
               STORE_FAST               0 (e)
               DELETE_FAST              0 (e)
```

?

Comment (albertz) [reply Python/bytecodes.c]: > Can we make POP_EXCEPT clear the locals

Yes, I also thought about that. However, then I thought, maybe there are other situations where the lifetime of objects is unexpected? Basically, whenever there is a `DELETE_FAST` on some local variable, and `f_locals` was accessed before, it would result in this maybe unexpected behavior that the object stays alive until `f_locals` is accessed again, or until the function finishes.

So, you say, this is more important now for leaving `except` blocks, and less important for other cases of `DELETE_FAST`?

Comment (markshannon) [reply Python/bytecodes.c]: The semantics of `f_locals` is somewhat broken, see PEP 667. Fiddling with `DELETE_FAST` is not going to fix that unfortunately, but is likely to break other stuff.

Which of the test cases fail with just the change to `frame_clear` to delete `f_locals`?

Comment (albertz) [reply Python/bytecodes.c]: > Which of the test cases fail with just the change to `frame_clear` to delete `f_locals`?

`test_locals_cleared_after_exception_handled` still fails with only the `frame_clear` change but not the `DELETE_FAST` change.

> The semantics of `f_locals` is somewhat broken, see PEP-667.

Ah, yes. I was thinking the same when looking through the code, that this `f_locals` vs fast locals (and those sync functions) potentially has a lot of issues.

> Fiddling with `DELETE_FAST` is not going to fix that unfortunately,

Well, it fixes the issue here, that there are still references to the object while there should not.

> but is likely to break other stuff.

I'm not sure. Why? If `DELETE_FAST` executes but `f_locals` is not touched by the op, then in any case `f_locals` is in a wrong state (not in sync with the fast locals anymore). So this change seems to be more safe to me than not doing it.

Or do you want to allow that bytecode can lead to such inconsistent state, and only Python source code is required to give consistent behavior?

Comment (gvanrossum) [reply Python/bytecodes.c]: @albertz, I appreciate that you've come this far in tracking down your issue. But I don't think the solution is to change bytecodes. Surely if we change `DELETE_FAST`, which implements `del x`, we should also change `STORE_FAST`, which implements `x = None` and also loses a reference to whatever was originally in `x`. But we don't want to make `STORE_FAST` slower, not even by one memory read + a conditional jump.

I also don't want bytecodes to mess with `f_locals` because the latter logically belongs to a totally different part of the interpreter.

I really think that the way to fix this, eventually, is to change the was `locals()` and `frame.f_locals` are implemented. There are two PEPs about this, 667 and 558. My preference would be 667, but both are currently blocked, mostly on people's time.

Comment (albertz) [reply Python/bytecodes.c]: Thanks for the response. Yes, I agree, PEP-667 would be much cleaner, and also solve this problem.

Instead of messing with bytecodes, do you have maybe another idea for a simple workaround until PEP-667 is implemented?

Or otherwise, for this PR here, should we remove the `DELETE_FAST` change, but still do the `frame.clear()` change? I think the change in `frame.clear()` should be fine, right?

In that case, remove also the `test_locals_cleared_after_exception_handled`, which would still fail then, or disable this test somehow?

**Comment by markshannon** [comment]
> Or otherwise, for this PR here, should we remove the DELETE_FAST change, but still do the frame.clear() change? I think the change in frame.clear() should be fine, right?

Yes, I think that's the way to go.

**Comment by albertz** [comment]
What about the test case? Should I also delete it? For reference, this one:
```python
class LocalsTest(unittest.TestCase):
    """
    Tests for locals.
    """

    def test_locals_cleared_after_exception_handled(self):
        # see gh-113939
        class C:
            pass
        wr = None
        def inner():
            nonlocal wr
            c = C()
            wr = weakref.ref(c)
            1/0
        try:
            inner()
        except ZeroDivisionError as exc:
            support.gc_collect()
            self.assertIsNotNone(wr())
            print(exc.__traceback__.tb_frame.f_locals)
        support.gc_collect()
        self.assertIsNone(wr())
```

The other test case (`test_clear_locals_after_f_locals_access`) can stay, of course.

**Comment by albertz** [comment]
> > Or otherwise, for this PR here, should we remove the DELETE_FAST change, but still do the frame.clear() change? I think the change in frame.clear() should be fine, right?
> 
> Yes, I think that's the way to go.

I did that now. So, PR can be merged now?

File: Lib/test/test_frame.py [hunk 5]
```diff
@@ -55,6 +55,26 @@ class C:
...
+    def test_clear_locals_after_f_locals_access(self):
+        # see gh-113939
+        class C:
+            pass
+        wr = None
+        def inner():
+            nonlocal wr
+            c = C()
+            wr = weakref.ref(c)
+            1/0
```
Comment (iritkatriel) [inline Lib/test/test_frame.py:68]:
```suggestion
            1/0

```

File: Lib/test/test_frame.py [hunk 6]
```diff
@@ -55,6 +55,26 @@ class C:
         # The reference was released by .clear()
         self.assertIs(None, wr())
 
+    def test_clear_locals_after_f_locals_access(self):
+        # see gh-113939
+        class C:
+            pass
```
Comment (iritkatriel) [inline Lib/test/test_frame.py:61]:
```suggestion
            pass

```

**Review by iritkatriel — APPROVED** [review]
Add whitespace, otherwise LGTM.
